File: OCProject/Facade/UICategory/UIButton+UU.m
//
//  UIButton+UU.m
//  OCProject
//
//  Created by Pan on 2020/5/13.
//  Copyright © 2020 xyz. All rights reserved.
//

#import "UIButton+UU.h"

@implementation UIButton (UU)

// 重新设置 image 及 title 的位置及间距
- (void)resetImageTitlePosition:(ButtonImageTitlePosition)position space:(CGFloat)space {    
    CGFloat imageWith = self.imageView.frame.size.width;
    CGFloat imageHeight = self.imageView.frame.size.height;
    
    CGFloat titleWidth = 0.0;
    CGFloat titleHeight = 0.0;
    if ([UIDevice currentDevice].systemVersion.floatValue >= 8.0) {
        // 由于iOS8中titleLabel的size为0，用下面的这种设置
        titleWidth = self.titleLabel.intrinsicContentSize.width;
        titleHeight = self.titleLabel.intrinsicContentSize.height;
    } else {
        titleWidth = self.titleLabel.frame.size.width;
        titleHeight = self.titleLabel.frame.size.height;
    }
    
    switch (position) {
        case ButtonImageTopTitleDown: {
            self.imageEdgeInsets = UIEdgeInsetsMake(-titleHeight-space, 0, 0, -titleWidth);
            self.titleEdgeInsets = UIEdgeInsetsMake(0, -imageWith, -imageHeight-space, 0);
        } break;
        case ButtonImageLeftTitleRight: {
            self.imageEdgeInsets = UIEdgeInsetsMake(0, -space/2.0, 0, space/2.0);
            self.titleEdgeInsets = UIEdgeInsetsMake(4, space/2.0, 0, -space/2.0);
        } break;
        case ButtonImageDownTitleTop: {
            self.imageEdgeInsets = UIEdgeInsetsMake(0, 0, -titleHeight-space, -titleWidth);
            self.titleEdgeInsets = UIEdgeInsetsMake(-imageHeight-space, -imageWith, 0, 0);
        } break;
        case ButtonImageRightTitleLeft: {
            self.imageEdgeInsets = UIEdgeInsetsMake(0, titleWidth+space/2.0, 0, -titleWidth-space/2.0);
            self.titleEdgeInsets = UIEdgeInsetsMake(0, -imageWith-space/2.0, 0, imageWith+space/2.0);
        } break;
    }
}

@end

File: SwiftProject/Facade/Extension/UIButton+UU.swift
//
//  UIButton+UU.swift
//  SwiftProject
//
//  Created by Pan on 2020/9/3.
//  Copyright © 2020 xyz. All rights reserved.
//

import UIKit

extension UIButton {
    
    /// 位置枚举
    enum ImageTitlePosition {
        /// image在上，title在下
        case imageTopTitleDown
        /// image在下，title在上
        case imageDownTitleTop
        /// image在左，title在右
        case imageLeftTitleRight
        /// image在右，title在左
        case imageRightTitleLeft
    }
    
    /// 重新设置 image 及 title 的位置及间距
    func resetImageTitlePosition(_ position: ImageTitlePosition, space: CGFloat) {
        let imageWith = imageView?.frame.size.width ?? 0.0
        let imageHeight = imageView?.frame.size.height ?? 0.0
        let titleWidth = titleLabel?.frame.size.width ?? 0.0
        let titleHeight = titleLabel?.frame.size.height ?? 0.0
        
        // 近图表或标题，则仍居中
        if imageWith == 0 || titleWidth == 0 { return }
                
        switch position {
        case .imageTopTitleDown:
            imageEdgeInsets = UIEdgeInsets(top: -titleHeight-space, left: titleWidth/2, bottom: 0, right: -titleWidth/2)
            titleEdgeInsets = UIEdgeInsets(top: 0, left: -imageWith, bottom: -imageHeight-space, right: 0)
        case .imageLeftTitleRight:
            imageEdgeInsets = UIEdgeInsets(top: 0, left: -space/2.0, bottom: 0, right: space/2.0)
            titleEdgeInsets = UIEdgeInsets(top: 0, left: space/2.0, bottom: 0, right: -space/2.0)
        case .imageDownTitleTop:
            imageEdgeInsets = UIEdgeInsets(top: 0, left: 0, bottom: -titleHeight-space, right: -titleWidth)
            titleEdgeInsets = UIEdgeInsets(top: -imageHeight-space, left: -imageWith, bottom: 0, right: 0)
        case .imageRightTitleLeft:
            imageEdgeInsets = UIEdgeInsets(top: 0, left: titleWidth+space/2.0, bottom: 0, right: -titleWidth-space/2.0)
            titleEdgeInsets = UIEdgeInsets(top: 0, left: -imageWith-space/2.0, bottom: 0, right: imageWith+space/2.0)
        }
    }
    
}
